### ISSUE TITLE
Build failed on ubuntu 14.04 LTS (dependecy sodium)

### ISSUE THREAD

**Issue by lcantat** [tezos/tezos#8]
please find bellow the information regarding this failure 

Step performed : 
opam switch tezos --alias-of 4.03.0 : OK 
eval `opam config env` : OK
make build-deps : Failure
> 
> [ERROR] The compilation of sodium failed at "make".
> Processing  1/2: [sodium: ocamlfind remove]
> #=== ERROR while installing sodium.0.5.0 ======================================#
> opam-version 1.2.2
> os           linux
> command      make
> path         /home/loic/.opam/tezos/build/sodium.0.5.0
> compiler     4.03.0
> exit-code    2
> env-file     /home/loic/.opam/tezos/build/sodium.0.5.0/sodium-12626-05296d.env
> stdout-file  /home/loic/.opam/tezos/build/sodium.0.5.0/sodium-12626-05296d.out
> stderr-file  /home/loic/.opam/tezos/build/sodium.0.5.0/sodium-12626-05296d.err
> [...]
> lib/sodium_stubs.c: In function ‘caml__74_crypto_pwhash_argon2i_str’:
> lib/sodium_stubs.c:646:4: error: implicit declaration of function ‘crypto_pwhash_argon2i_str’ [-Werror=implicit-function-declaration]
>     int x490 = crypto_pwhash_argon2i_str(x479, x480, x481, x484, x487);
>      ^
>  lib/sodium_stubs.c: In function ‘caml__75_crypto_pwhash_argon2i_str_verify’:
>  lib/sodium_stubs.c:655:4: error: implicit declaration of function ‘crypto_pwhash_argon2i_str_verify’ [-Werror=implicit-function-declaration]
>      int x499 = crypto_pwhash_argon2i_str_verify(x494, x495, x496);
>      ^
>  cc1: all warnings being treated as errors
>  Command exited with code 2.
>  make: *** [all] Erreur 10
> =-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
> The following actions were aborted
>   ∗  install tezos-deps dev
> The following actions failed
>   ∗  install sodium 0.5.0
> No changes have been performed
> 
as recommended I try `opam depext sodium.0.5.0`
>  Detecting depexts using flags: x86_64 linux ubuntu
>  The following system packages are needed:
>   - libsodium-dev
>   - m4
>   - pkg-config
>  All required OS packages found.`

In order to satisfy libsodium-dev dependency I add to my ubunty apt repository the following dependency : 
> deb http://ppa.launchpad.net/chris-lea/libsodium/ubuntu trusty main 
> deb-src http://ppa.launchpad.net/chris-lea/libsodium/ubuntu trusty main `

**Comment by eliyastein** [comment]
I'm also trying to do a build on Ubuntu that's failing in the same spot:
```=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
Processing  1/2: [sodium: make]
[ERROR] The compilation of sodium failed at "make".
Processing  1/2: [sodium: ocamlfind remove]
#=== ERROR while installing sodium.dev ========================================#
# opam-version 1.2.2
# os           linux
# command      make
# path         /home/chinaski/.opam/tezos/build/sodium.dev
# compiler     4.04.0
# exit-code    2
# env-file     /home/chinaski/.opam/tezos/build/sodium.dev/sodium-16113-e7966f.env
# stdout-file  /home/chinaski/.opam/tezos/build/sodium.dev/sodium-16113-e7966f.out
# stderr-file  /home/chinaski/.opam/tezos/build/sodium.dev/sodium-16113-e7966f.err
### stdout ###
# [...]
# lib/sodium_stubs.c:646:15: error: implicit declaration of function ‘crypto_pwhash_argon2i_str’ [-Werror=implicit-function-declaration]
#     int x490 = crypto_pwhash_argon2i_str(x479, x480, x481, x484, x487);
#                ^
# lib/sodium_stubs.c: In function ‘caml__75_crypto_pwhash_argon2i_str_verify’:
# lib/sodium_stubs.c:655:15: error: implicit declaration of function ‘crypto_pwhash_argon2i_str_verify’ [-Werror=implicit-function-declaration]
#     int x499 = crypto_pwhash_argon2i_str_verify(x494, x495, x496);
#                ^
# cc1: all warnings being treated as errors
# Command exited with code 2.
# Makefile:13: recipe for target 'all' failed
### stderr ###
# make: *** [all] Error 10



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install tezos-deps dev
The following actions failed
  ∗  install sodium dev
No changes have been performed

=-=- sodium.dev troobleshooting -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=> This package relies on external (system) dependencies that may be missing.
   `opam depext sodium.dev' may help you find the correct installation for
   your system.
Makefile:18: recipe for target 'build-deps' failed
make: *** [build-deps] Error 4
```

**Comment by tezos** [comment]
You need a more recent version of libsodium. Try 
1.0.11 from https ://launchpad.net/ubuntu/+source/libsodium

**Comment by eliyastein** [comment]
Thanks @tezos I appreciate the guidance. I installed libsodium 1.011 and that got me through 'make build-deps', but now it fails during make (see below). I'm assuming I installed libsodium incorrectly? Would love your thoughts: 

```findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/chinaski/.opam/tezos/lib/ocaml/compiler-libs, /home/chinaski/.opam/tezos/lib/ocaml
LINK compiler.cmxa
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/chinaski/.opam/tezos/lib/ocaml/compiler-libs, /home/chinaski/.opam/tezos/lib/ocaml
OCAMLOPT (compiler.cmxa) compiler_main.cmx
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/chinaski/.opam/tezos/lib/ocaml/compiler-libs, /home/chinaski/.opam/tezos/lib/ocaml
LINK tezos-protocol-compiler
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/chinaski/.opam/tezos/lib/ocaml/compiler-libs, /home/chinaski/.opam/tezos/lib/ocaml
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__29_crypto_pwhash_argon2i_saltbytes':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:181: undefined reference to `crypto_pwhash_argon2i_saltbytes'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `_crypto_pwhash_argon2i_strbyte::caml(short)':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:186: undefined reference to `crypto_pwhash_argon2i_strbytes'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__31_crypto_pwhash_argon2i_memlimit_interactive':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:191: undefined reference to `crypto_pwhash_argon2i_memlimit_interactive'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__32_crypto_pwhash_argon2i_memlimit_moderate':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:196: undefined reference to `crypto_pwhash_argon2i_memlimit_moderate'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__33_crypto_pwhash_argon2i_memlimit_sensitive':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:201: undefined reference to `crypto_pwhash_argon2i_memlimit_sensitive'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__34_crypto_pwhash_argon2i_opslimit_interactive':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:206: undefined reference to `crypto_pwhash_argon2i_opslimit_interactive'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__35_crypto_pwhash_argon2i_opslimit_moderate':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:211: undefined reference to `crypto_pwhash_argon2i_opslimit_moderate'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__36_crypto_pwhash_argon2i_opslimit_sensitive':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:216: undefined reference to `crypto_pwhash_argon2i_opslimit_sensitive'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__37_crypto_pwhash_argon2i_alg_argon2i13':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:221: undefined reference to `crypto_pwhash_argon2i_alg_argon2i13'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__38_crypto_pwhash_argon2i':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:236: undefined reference to `crypto_pwhash_argon2i'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__74_crypto_pwhash_argon2i_str':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:646: undefined reference to `crypto_pwhash_argon2i_str'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__75_crypto_pwhash_argon2i_str_verify':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:655: undefined reference to `crypto_pwhash_argon2i_str_verify'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__76_crypto_pwhash_argon2i_str':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:666: undefined reference to `crypto_pwhash_argon2i_str'
/home/chinaski/.opam/tezos/lib/sodium/libsodium_stubs.a(sodium_stubs.o): In function `caml__77_crypto_pwhash_argon2i_str_verify':
/home/chinaski/.opam/tezos/build/sodium.dev/_build/lib/sodium_stubs.c:675: undefined reference to `crypto_pwhash_argon2i_str_verify'
collect2: error: ld returned 1 exit status
File "caml_startup", line 1:
Error: Error during linking
Makefile:175: recipe for target '../tezos-protocol-compiler' failed
make[1]: *** [../tezos-protocol-compiler] Error 2
rm compiler/embedded_cmis.ml
make[1]: Leaving directory '/home/chinaski/tezos/src'
Makefile:3: recipe for target 'all' failed
make: *** [all] Error 2```

**Comment by tezos** [comment]
Did your run ldconfig? It might be easier to help you troubleshoot live. Can you join our slack channel and ping me (@arthurb)

**Comment by OCamlPro-Henry** [comment]
@eliyastein What about: `opam reinstall sodium; make clean; make` ?

**Comment by itsnotlupus** [comment]
Bumped into a similar scenario on Ubuntu 16.04, which can only get libsodium 1.0.8-5 through official channels. After reading this page, I ended up grabbing packages built for a different release, because that never backfires. (https://launchpad.net/ubuntu/zesty/amd64/libsodium18/1.0.11-1 and https://launchpad.net/ubuntu/zesty/amd64/libsodium-dev/1.0.11-1 )

One `opam reinstall sodium; make clean; make` later, and I'm staring at an `Error: Unbound module Ssl` message, which seems to hint something's still not quite happy, but at least I'm getting a little bit further than @eliyastein did.
```
LINK embedded_proto_demo.cmxa
OCAMLOPT (embedded_proto_genesis.cmxa) local_environment.cmx
OCAMLOPT (embedded_proto_genesis.cmxa) data.cmx
OCAMLOPT (embedded_proto_genesis.cmxa) services.cmx
OCAMLOPT (embedded_proto_genesis.cmxa) main.cmx
PACK (embedded_proto_genesis.cmxa) embedded_proto_genesis.cmx
OCAMLOPT register_Embedded_proto_genesis.cmx
LINK embedded_proto_genesis.cmxa
OCAMLOPT (node.cmxa) node_identity_file.cmi
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_identity_file.cmx
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_config_file.cmi
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_config_file.cmx
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_shared_arg.cmi
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_shared_arg.cmx
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_run_command.cmi
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
OCAMLOPT (node.cmxa) node_run_command.cmx
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/itsnotlupus/.opam/tezos/lib/ocaml/compiler-libs, /home/itsnotlupus/.opam/tezos/lib/ocaml
File "node/main/node_run_command.ml", line 128, characters 6-20:
Error: Unbound module Ssl
Makefile:301: recipe for target 'node/main/node_run_command.cmx' failed
make[1]: *** [node/main/node_run_command.cmx] Error 2
rm compiler/embedded_cmis.ml
make[1]: Leaving directory '/store/dev/tezos/tezos/src'
Makefile:3: recipe for target 'all' failed
make: *** [all] Error 2
```

I'll keep poking at this, but I'd welcome any slightly cryptic hints in the meanwhile.

**Comment by tezos** [comment]
You should amend src/Makefile.files, around line 64, replace lwt by lwt ssl 
Sorry, tech preparations for the ico are keeping us very busy but we'll push a fish to the makefile soon.

**Comment by itsnotlupus** [comment]
@tezos thank you, that did it! 
It looks like that's all that's missing to compile some working binaries.

**Comment by LRonHubs** [comment]
@itsnotlupus, how did you get ocaml 4.04 to install on ubuntu 16?

**Comment by tezos** [comment]
opam switch

**Comment by itsnotlupus** [comment]
@LRonHubs not entirely sure. I think I already had it installed from an earlier thing.  
It looks like you can `apt install` ocaml and opam straight from the standard ubuntu repos.    
While that'll ostensibly only install ocaml 4.02, that `opam switch` command given in this repo's readme will apparently grab 4.04 and allow the rest of the build process to go on happily.

**Comment by LRonHubs** [comment]
i'm working through this with a friend. the opam switch command completes. but when we run "make build-deps", the ocaml library immediately uses 4.02. Is there a way we're supposed to specify the alias we just installed to? we're still in the same directory and everything. this is both of our first times using ocaml; both running ubuntu 16.04; one processor is 32 bit the other is 64 bit

**Comment by LRonHubs** [comment]
Okay the "eval `opam config env`" command was the problem.

In  order to get this to work on ubuntu, you need to add the output of this command to the ~/.bashrc file. 
You should be able to run each command individually, and i think the eval command is supposed to do this, but the eval function didn't work for me. adding it to the bashrc is not that bad and worked for me.

**Comment by itsnotlupus** [comment]
interesting. the instructions worked for me without any extra tweaks.  
I'm not sure what's different between our setups to account for that. A different shell maybe?  
Be careful if you keep those lines in your .bashrc, it's probably going to mess with something in the future (due to its flat overriding of PATH without referencing $PATH.)

**Comment by LRonHubs** [comment]
Right i did make those changes actually thanks.

I just want to add that to actually use the modern version of sodium you have to change line 49 in install_build_deps.sh to "opam pin --yes add --no-action  sodium https://launchpad.net/ubuntu/+source/libsodium". then you can run "make build-deps; make;" and get everything to work. 

how is opam able to install these packages when i haven't used sudo at all at any point in this process? i.e., "Opam will use sudo to install the following packages:"; how is this possible when i'm not in sudo?

**Comment by OCamlPro-Iguernlala** [comment]
If you are not a sudoer, OPAM will (probably) not be able to install external (system) dependencies. You should ask a sudoer to install them for you

**Comment by jplevyak** [comment]
on 16.04 LTS I ended up having to install libsodium version 1.0.11-2 manually:

libsodium18_1.0.11-2_amd64.deb
libsodium-dev_1.0.11-2_amd64.deb

and openssl version 1.1.0

and then setting

EXTRA_OCAMLFLAGS "-ccopt -L/usr/local/lib"

before running make

And then unsetting it (as one of the later targets did not understand -ccopt) and completing
build by running "make" again.

**Comment by LRonHubs** [comment]
Change line 63 in Makefile.files from "lwt" to lwt ssl
nano +63 /src/Makefile.files; 

On amd64: 


wget http://launchpadlibrarian.net/283262471/libsodium18_1.0.11-1_amd64.deb; wget http://launchpadlibrarian.net/283262467/libsodium-dev_1.0.11-1_amd64.deb; sudo dpkg -i *.deb; sudo apt-get -f install; make build-deps; make;


On i386:

 wget http://launchpadlibrarian.net/283262444/libsodium-dev_1.0.11-1_i386.deb; wget http://launchpadlibrarian.net/283262445/libsodium18_1.0.11-1_i386.deb; sudo dpkg -i *.deb; sudo apt-get -f install; make build-deps; make;

You might get an error saying that i386 is not your operating system. that is not a big problem just text me

**Comment by dantrevino** [comment]
None of this is working for me.  I modified src/Makefile.files to "lwt ssl", updated libsodium.  make clean; make build-deps.  I'm getting git/gpg errors:

tezos master ✗ 10d △ ◒ ➜ make build-deps
```
+ true
+ opam pin --yes remove --no-action --dev-repo ocplib-resto
[NOTE] ocplib-resto is not pinned.
+ opam pin --yes add --no-action --dev-repo sodium
[NOTE] Package sodium is already git-pinned to
       https://github.com/dsheets/ocaml-sodium.git.
       This will erase any previous custom definition.
Proceed ? [Y/n] y

[ERROR] Command "git commit --allow-empty -m opam-git-init" failed:
# opam-version 1.2.2
# os           linux
# command      git commit --allow-empty -m opam-git-init
# path         /home/dan/.opam/tezos/packages.dev/sodium
# exit-code    128
# env-file     /home/dan/.opam/log/log-17118-ccabea.env
# stdout-file  /home/dan/.opam/log/log-17118-ccabea.out
# stderr-file  /home/dan/.opam/log/log-17118-ccabea.err
### stderr ###
# gpg: skipped "Opam <opam@ocaml.org>": No secret key
# gpg: signing failed: No secret key
# error: gpg failed to sign the data
# fatal: failed to write commit object

Makefile:18: recipe for target 'build-deps' failed
make: *** [build-deps] Error 1
```
tezos master ✗ 10d △ ◒ ⍉ ➜ cat /home/dan/.opam/log/log-17118-ccabea.err
```
gpg: skipped "Opam <opam@ocaml.org>": No secret key
gpg: signing failed: No secret key
error: gpg failed to sign the data
fatal: failed to write commit object
```

**Comment by hal9000xp** [comment]
I successfully built Tezos under Ubuntu 14.04. Here what I did:

1. Downloaded:

https://download.libsodium.org/libsodium/releases/libsodium-1.0.13.tar.gz

2. Built and installed using instruction:

https://download.libsodium.org/doc/installation/

3. Made and installed **fake** package for **libsodium-dev** so that tezos build-deps thinks there is a package installed in the system while I use custom build for sodium (step 1,2):

```
sudo apt-get install equivs
equivs-control libsodium-dev
equivs-build libsodium-dev
sudo dpkg -i libsodium-dev_1.0_all.deb
```

After that I successfully ran:

```
make build-deps
make
```
